driver/src/price_finding/naive_solver.rs

```diff
@@ -497,12 +472,110 @@ pub mod tests {
                 buy_amount: 19960,
             },
         ];
+        let state = create_account_state_with_balance_for(&orders);
 
-        let mut solver = NaiveSolver::new(Some(Fee {
+        let fee = Some(Fee {
             token: 2,
             percentage: 0.001,
-        }));
+        });
+        let mut solver = NaiveSolver::new(fee.clone());
         let res = solver.find_prices(&orders, &state).unwrap();
         assert_eq!(res.executed_sell_amounts, [0, 0]);
+
+        check_solution(&orders, res, &fee).unwrap();
+    }
+
+    fn order_pair_first_fully_matching_second() -> Vec<Order> {
+        vec![
+            Order {
+                batch_information: None,
+                account_id: H160::from(1),
+                sell_token: 0,
+                buy_token: 1,
+                sell_amount: 52 * 10u128.pow(18),
+                buy_amount: 4 * 10u128.pow(18),
+            },
+            Order {
+                batch_information: None,
+                account_id: H160::from(0),
+                sell_token: 1,
+                buy_token: 0,
+                sell_amount: 15 * 10u128.pow(18),
+                buy_amount: 180 * 10u128.pow(18),
+            },
+        ]
+    }
+
+    fn order_pair_both_fully_matched() -> Vec<Order> {
+        vec![
+            Order {
+                batch_information: None,
+                account_id: H160::from(1),
+                sell_token: 2,
+                buy_token: 1,
+                sell_amount: 10 * 10u128.pow(18),
+                buy_amount: 10 * 10u128.pow(18),
+            },
+            Order {
+                batch_information: None,
+                account_id: H160::from(1),
+                sell_token: 1,
+                buy_token: 2,
+                sell_amount: 16 * 10u128.pow(18),
+                buy_amount: 8 * 10u128.pow(18),
+            },
+        ]
+    }
+
+    fn check_solution(
+        orders: &[Order],
+        solution: Solution,
+        fee: &Option<Fee>,
+    ) -> Result<(), String> {
+        let mut token_conservation = HashMap::new();
+        for (i, order) in orders.iter().enumerate() {
+            let buy_token_price = solution.prices[order.buy_token as usize];
+            let sell_token_price = solution.prices[order.sell_token as usize];
+
+            let exec_sell_amount = solution.executed_sell_amounts[i];
+            let mut exec_buy_amount = if buy_token_price > 0 {
+                exec_sell_amount * sell_token_price / buy_token_price
+            } else {
+                0
+            };
+            if let Some(fee) = fee {
+                exec_buy_amount = (exec_buy_amount as f64 * (1.0 - fee.percentage)) as u128;
```

Comment:
nit: you might wanna use the same fee calculation logic with the same rounding issues as in the contract.